### PR TITLE
object counting in video using yolov8

### PR DESCRIPTION
**Pull request by prince0310**
lets make life more easier
#### CLI
YOLOv8 may be used directly in the Command Line Interface (CLI) with a `yolo` command:

bash

`yolo predict model source="video.mp4" classes count = True`

YOLOv8 may be used directly in the Command Line Interface (CLI) with a yolo command:

`yolo predict model source="video.mp4" classes= 0 count = True`

#### Python

YOLOv8 may also be used directly in a Python environment, and accepts the
same [arguments](https://docs.ultralytics.com/cfg/) as in the CLI example above:

```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

# Use the model
results = model.predict(source="video.mp4" classes= 0  count = True)  


<!--
copilot:all
-->
### <samp>🤖 Generated by Copilot at 2f9cf30</samp>

### Summary
:sparkles::wrench::bar_chart:

<!--
1.  :sparkles: This emoji is often used to indicate the addition of a new feature or enhancement to a project. In this case, it could represent the new configuration option `count` and the new counting functionality in the `predict.py` file.
2.  :wrench: This emoji is often used to indicate the adjustment or improvement of an existing feature or functionality. In this case, it could represent the modification of the `non_max_suppression` function and the `box_label` function to support the counting feature.
3.  :bar_chart: This emoji is often used to indicate the display or analysis of some data or statistics. In this case, it could represent the counting of the number of objects of each class and the visualization of the results on the output image.
-->
This pull request implements a new feature to count the number of detected objects of each class in a region of interest on the output image. It modifies the `plotting.py`, `predict.py`, `ops.py` files and the `DetectionPredictor` class, and adds a new configuration option `count` to the `default.yaml` file and the `BasePredictor` class.

> _`count` option added_
> _detect and plot objects in_
> _autumn images_

### Walkthrough
*  Add a new option `count` to enable or disable counting the number of detected objects in the output image ([link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-9d692d4ad15d420a92105fcd432fb1d2221b71d5af055bea3bd2b70523ce6743R65), [link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-bd2831d027341da2474db760605211ea00b347396b0d708a4343a2b709eb3675R89), [link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-d47b2bd99d71bef47ccb82b27f8a102c04b98072070f94d0ba341fa3221f8b81L11-R17), [link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-d47b2bd99d71bef47ccb82b27f8a102c04b98072070f94d0ba341fa3221f8b81L23-R28), [link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-d47b2bd99d71bef47ccb82b27f8a102c04b98072070f94d0ba341fa3221f8b81L29-R35), [link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-d47b2bd99d71bef47ccb82b27f8a102c04b98072070f94d0ba341fa3221f8b81R42))
*  Modify the `non_max_suppression` function in `ops.py` to accept the `count` parameter and pass it to the `DetectionPredictor` class ([link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-6ab52db19716f4198ed847132149ab432f032ed39c99c264a9de386446bf7f88R137))
*  Update the `Annotator` class in `plotting.py` to store and return the center coordinates of the bounding boxes for counting purposes ([link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-7e75b3f968b9868ef6f0a86148cf6c16a5224e7fdd45f5c2eb5e9bc0fc744756R44-R45), [link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-7e75b3f968b9868ef6f0a86148cf6c16a5224e7fdd45f5c2eb5e9bc0fc744756R87-R94), [link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-7e75b3f968b9868ef6f0a86148cf6c16a5224e7fdd45f5c2eb5e9bc0fc744756L150-R164))
*  Implement the counting logic in the `write_results` method of the `DetectionPredictor` class in `predict.py` using the `cv2` module and the `ans` method of the `Annotator` class ([link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-d47b2bd99d71bef47ccb82b27f8a102c04b98072070f94d0ba341fa3221f8b81R4), [link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-d47b2bd99d71bef47ccb82b27f8a102c04b98072070f94d0ba341fa3221f8b81L54), [link](https://github.com/ultralytics/ultralytics/pull/575/files?diff=unified&w=0#diff-d47b2bd99d71bef47ccb82b27f8a102c04b98072070f94d0ba341fa3221f8b81R66-R81))



## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Implemented an object counting feature in the YOLO detection system.

### 📊 Key Changes
- Added a `count` parameter to various configuration and function signatures to enable object counting.
- Introduced code in `plotting.py` to draw a threshold line, detect the center of bounding boxes, and maintain a count.
- In `predict.py`, added functionality to increment and display the object count on the image if the count parameter is enabled.

### 🎯 Purpose & Impact
- **Purpose**: To provide users with the ability to not only detect objects but also to count the occurrences of each object type in an image.
- **Impact**: Users can now get real-time counts of objects which could be particularly useful in scenarios like traffic monitoring, retail analytics, or any other domain needing quantification of items or entities.

🚦 With this update, the Ultralytics YOLO system extends its capabilities from pure detection to also serve as a counting tool, offering richer analytics for a variety of applications.

### CODE REVIEW

**Comment by AyushExel** [comment]
Thanks for the PR!
@glenn-jocher this seems like a good idea. Let's refactor a bit and merge this. What do you think?

**Comment by prince0310** [comment]
@AyushExel  @glenn-jocher can this be merged ?

**Comment by AyushExel** [comment]
I'll take over this PR and add some minor updates this weekend.

**Comment by glenn-jocher** [comment]
@prince0310 @AyushExel I think we should rely on our new track mode to build a count feature.

**Comment by SOUMIA21** [comment]
Do you have any idea how to control the frames of the video? 
For example, writing on the frame to account for something. and predicting objects using Yolov8 at the same time.

**Comment by glenn-jocher** [comment]
@SOUMIA21 to control the frames of the video and also use YOLOv8 for object detection, one approach can be to extract the frames from the video, process them in a loop to apply the required modifications with the help of OpenCV library or any other image processing library of your choice, and then pass each frame to YOLOv8 for object detection. Then, the modified frames with bounding boxes and labels generated by YOLOv8 can be combined to generate an output video. To optimize the process, we can pre-cache frames and re-use them if there are no changes compared to previous frames. Additionally, parallel processing can be used to speed up the task.

**Comment by emi-net21** [comment]
@prince0310 Do you have any screenshot to see how is the result?

**Comment by glenn-jocher** [comment]
@emi-net21 sure! Unfortunately, I cannot see or provide screenshots within this GitHub issue since it is a text-based platform. However, you can try running YOLOv8 on your own images or videos to see the results. YOLOv8 is a state-of-the-art object detection algorithm that can accurately detect and localize objects in images and videos.

To use YOLOv8, you need to have the appropriate weights file for the pre-trained model, as well as the configuration file. With these files, you can run the YOLOv8 algorithm on your input data. It will output the bounding boxes, labels, and confidence scores of the detected objects.

If you encounter any issues or have any specific questions about using YOLOv8, feel free to ask, and we'll be happy to assist you!
